File: implementations/cpp/game.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

void write_state(char b[3][3], string turn, string winner, string existing_log, string new_move) {
    ofstream f("current_state.json");
    f << "{\n  \"board\": [\n";
    for(int i=0; i<3; i++) {
        f << "    [\"" << (b[i][0]?string(1,b[i][0]):"") << "\", \"" << (b[i][1]?string(1,b[i][1]):"") << "\", \"" << (b[i][2]?string(1,b[i][2]):"") << "\"]" << (i==2?"":",") << "\n";
    }
    string log_content = existing_log;
    if (new_move != "") {
        if (log_content.find("{") != string::npos) log_content += ", ";
        log_content += new_move;
    }
    f << "  ],\n  \"turn\": \"" << turn << "\",\n  \"winner\": " << (winner==""?"null":"\""+winner+"\"") << ",\n  \"log\": [" << log_content << "]\n}";
}

int main() {
    char b[3][3] = {0};
    string turn = "X";
    string winnerStr = "null";
    char* action_env = getenv("ACTION");
    char* cell_env = getenv("CELL");
    string action = action_env ? action_env : "put";
    string cell = cell_env ? cell_env : "";

    ifstream f("current_state.json");
    string f_line, full_content;
    int r=0;
    if (f.is_open()) {
        while(getline(f, f_line)) {
            full_content += f_line + "\n";
            if(f_line.find("[") != string::npos && f_line.find("board") == string::npos && f_line.find("log") == string::npos) {
                if(f_line.find("\"X\"") != string::npos) b[r][0] = 'X'; else if(f_line.find("\"O\"") != string::npos) b[r][0] = 'O';
                size_t pos = f_line.find(",");
                if(pos != string::npos) {
                    if(f_line.find("\"X\"", pos) != string::npos) b[r][1] = 'X'; else if(f_line.find("\"O\"", pos) != string::npos) b[r][1] = 'O';
                    pos = f_line.find(",", pos+1);
                    if(pos != string::npos) {
                        if(f_line.find("\"X\"", pos) != string::npos) b[r][2] = 'X'; else if(f_line.find("\"O\"", pos) != string::npos) b[r][2] = 'O';
                    }
                }
                if(r<2) r++;
            }
            if(f_line.find("\"turn\": \"O\"") != string::npos) turn = "O";
            if(f_line.find("\"winner\":") != string::npos) {
                size_t p1 = f_line.find(":");
                size_t p2 = f_line.find_first_of("\",\n", p1+1);
                size_t p3 = f_line.find_last_of("\"", p2);
                winnerStr = f_line.substr(p1+1);
                if (winnerStr.find("null") != string::npos) winnerStr = "null";
                else if (winnerStr.find("X") != string::npos) winnerStr = "X";
                else if (winnerStr.find("O") != string::npos) winnerStr = "O";
                else if (winnerStr.find("draw") != string::npos) winnerStr = "draw";
            }
        }
        f.close();
    }

    if (winnerStr != "null") return 0;

    string existing_log = "";
    size_t log_start = full_content.find("\"log\": [");
    if(log_start != string::npos) {
        log_start += 8;
        size_t log_end = full_content.find("]", log_start);
        if(log_end != string::npos) {
            existing_log = full_content.substr(log_start, log_end - log_start);
            size_t first = existing_log.find_first_not_of(" \n\r\t");
            size_t last = existing_log.find_last_not_of(" \n\r\t");
            if (first != string::npos && last != string::npos) existing_log = existing_log.substr(first, last - first + 1);
            else existing_log = "";
        }
    }

    if(action == "reset") {
        char empty_b[3][3] = {0};
        write_state(empty_b, "X", "", "", "");
    }
    else if(cell != "") {
        int row = cell[1]-'1', col = cell[0]-'A';
        if(row>=0 && row<3 && col>=0 && col<3 && b[row][col]==0) {
            b[row][col] = turn[0];
            int lns[8][6] = {{0,0,0,1,0,2},{1,0,1,1,1,2},{2,0,2,1,2,2}, {0,0,1,0,2,0},{0,1,1,1,2,1},{0,2,1,2,2,2}, {0,0,1,1,2,2},{0,2,1,1,2,0}};
            bool win = false;
            for(int i=0; i<8; i++) if(b[lns[i][0]][lns[i][1]] && b[lns[i][0]][lns[i][1]]==b[lns[i][2]][lns[i][3]] && b[lns[i][2]][lns[i][3]]==b[lns[i][4]][lns[i][5]]) win=true;
            string new_move = "{\"player\": \"" + turn + "\", \"cell\": \"" + cell + "\"}";
            if(win) write_state(b, turn, turn, existing_log, new_move);
            else {
                bool full = true; for(int i=0; i<3; i++) for(int j=0; j<3; j++) if(b[i][j]==0) full=false;
                if(full) write_state(b, turn, "draw", existing_log, new_move);
                else write_state(b, (turn=="X"?"O":"X"), "", existing_log, new_move);
            }
        }
    }
    return 0;
}

File: implementations/kotlin/Game.kt
import java.io.File

data class Move(val player: String, val cell: String)
data class State(var board: List<MutableList<String>>, var turn: String, var winner: String?, var log: List<Move>)

fun main() {
    val file = File("current_state.json")
    val json = file.readText()
    val state = parseState(json)

    val cell = System.getenv("CELL")?.toUpperCase() ?: ""
    val action = System.getenv("ACTION") ?: "put"

    if (action == "reset") {
        val isFull = state.board.all { row -> row.all { it.isNotEmpty() } }
        if (state.winner != null || isFull) {
            state.board = listOf(mutableListOf("","",""), mutableListOf("","",""), mutableListOf("","",""))
            state.turn = "X"
            state.winner = null
            state.log = emptyList()
        }
    } else if (cell.isNotEmpty() && state.winner == null) {
        val r = cell[1] - '1'
        val c = cell[0] - 'A'
        if (r in 0..2 && c in 0..2 && state.board[r][c].isEmpty()) {
            state.board[r][c] = state.turn
            state.log = state.log + Move(state.turn, cell)
            
            val win = checkWinner(state.board)
            if (win != null) state.winner = win
            else {
                if (state.board.all { row -> row.all { it.isNotEmpty() } }) state.winner = "draw"
                else state.turn = if (state.turn == "X") "O" else "X"
            }
        }
    }
    file.writeText(stateToJson(state))
}

fun parseState(json: String): State {
    val board = mutableListOf(mutableListOf("","",""), mutableListOf("","",""), mutableListOf("","",""))
    var turn = "X"
    var winner: String? = null
    val log = mutableListOf<Move>()

    if (json.contains("\"board\"")) {
        val boardIdx = json.indexOf("\"board\"")
        val startIdx = json.indexOf("[", boardIdx)
        var row = 0
        var pos = startIdx + 1
        while (row < 3) {
            val rowStart = json.indexOf("[", pos)
            if (rowStart == -1) break
            val rowEnd = json.indexOf("]", rowStart)
            val rowStr = json.substring(rowStart + 1, rowEnd)
            val cells = rowStr.split(",")
            for (col in 0 until 3) {
                if (col < cells.size) {
                    board[row][col] = cells[col].trim().replace("\"", "")
                }
            }
            row++
            pos = rowEnd + 1
        }
    }
    
    if (json.contains("\"turn\"")) {
        val idx = json.indexOf("\"turn\"")
        val valStart = json.indexOf("\"", json.indexOf(":", idx))
        val valEnd = json.indexOf("\"", valStart + 1)
        turn = json.substring(valStart + 1, valEnd)
    }
    
    if (json.contains("\"winner\"")) {
        val idx = json.indexOf("\"winner\"")
        val colonIdx = json.indexOf(":", idx)
        val nextQuote = json.indexOf("\"", colonIdx)
        val nextComma = json.indexOf(",", colonIdx)
        val nextBrace = json.indexOf("}", colonIdx)
        val end = if (nextComma != -1) nextComma else nextBrace
        val value = json.substring(colonIdx + 1, end).trim()
        if (value.startsWith("\"")) {
            winner = value.substring(1, value.lastIndexOf("\""))
        } else if (value == "null") {
            winner = null
        } else {
            winner = value
        }
    }
    
    if (json.contains("\"log\"")) {
        val idx = json.indexOf("\"log\"")
        val start = json.indexOf("[", idx)
        val end = json.lastIndexOf("]")
        val logContent = json.substring(start + 1, end).trim()
        if (logContent.isNotEmpty()) {
            val entries = logContent.split("}")
            for (entry in entries) {
                if (entry.contains("{")) {
                    val pIdx = entry.indexOf("\"player\"")
                    val pValStart = entry.indexOf("\"", entry.indexOf(":", pIdx))
                    val pValEnd = entry.indexOf("\"", pValStart + 1)
                    val p = entry.substring(pValStart + 1, pValEnd)
                    
                    val cIdx = entry.indexOf("\"cell\"")
                    val cValStart = entry.indexOf("\"", entry.indexOf(":", cIdx))
                    val cValEnd = entry.indexOf("\"", cValStart + 1)
                    val c = entry.substring(cValStart + 1, cValEnd)
                    log.add(Move(p, c))
                }
            }
        }
    }

    return State(board, turn, winner, log)
}

fun stateToJson(s: State): String {
    val sb = StringBuilder()
    sb.append("{\n  \"board\": [\n")
    for (i in 0 until 3) {
        sb.append("    [\"${s.board[i][0]}\", \"${s.board[i][1]}\", \"${s.board[i][2]}\"]")
        if (i < 2) sb.append(",")
        sb.append("\n")
    }
    sb.append("  ],\n  \"turn\": \"${s.turn}\",\n")
    sb.append("  \"winner\": ").append(if (s.winner == null) "null" else "\"${s.winner}\"").append(",\n")
    sb.append("  \"log\": [\n")
    for (i in s.log.indices) {
        val m = s.log[i]
        sb.append("    {\"player\": \"${m.player}\", \"cell\": \"${m.cell}\"}")
        if (i < s.log.size - 1) sb.append(",")
        sb.append("\n")
    }
    sb.append("  ]\n}")
    return sb.toString()
}

fun checkWinner(b: List<List<String>>): String? {
    val lines = listOf(listOf(0,1,2),listOf(3,4,5),listOf(6,7,8),listOf(0,3,6),listOf(1,4,7),listOf(2,5,8),listOf(0,4,8),listOf(2,4,6))
    val flat = b.flatten()
    for (l in lines) {
        if (flat[l[0]].isNotEmpty() && flat[l[0]] == flat[l[1]] && flat[l[0]] == flat[l[2]]) return flat[l[0]]
    }
    return null
}

